cpp/src/stream_compaction/stream_compaction.cu

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf.h>
+#include <types.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <stream_compaction.hpp>
+#include <bitmask/bit_mask.cuh>
+#include <utilities/device_atomics.cuh>
+#include <utilities/cudf_utils.h>
+#include <utilities/error_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+#include <utilities/wrapper_types.hpp>
+#include <utilities/cuda_utils.hpp>
+#include <cub/cub.cuh>
+
+using bit_mask::bit_mask_t;
+
+namespace {
+
+static constexpr int warp_size = 32;
+
+// Returns true if the mask is true and valid (non-null) for index i
+__device__ inline 
+bool valid_and_true(cudf::bool8 const * __restrict__ data,
+                    bit_mask_t const * __restrict__ bitmask,
+                    gdf_index_type i)
+{
+  bool valid = bit_mask::is_valid(bitmask, i);
+  return (cudf::true_v == data[i]) && valid;
+}
+
+}  // namespace
+
+namespace cudf {
+
+// Compute the count of elements that pass the mask within each block
+template <int block_size, int per_thread>
+__global__ void compute_block_counts(gdf_size_type  * __restrict__ block_counts,
+                                     cudf::bool8 const* __restrict__ mask_data,
+                                     bit_mask_t const* __restrict__ mask_valid,
+                                     gdf_size_type mask_size)
+{
+  int tid = threadIdx.x + per_thread * block_size * blockIdx.x;
+  int count = 0;
+
+  for (int i = 0; i < per_thread; i++) {
```

Comment:
I tried block_reduce, but it was slower, because it doesn't optimize for boolean inputs. Since it has to __syncthreads() anyway, a __syncthreads_count() is pretty optimal.

I thought I experimented with unroll, but will do so again and report back.

Reply:
I tried #pragma unroll again and any amount of unrolling reduced performance.